### PR TITLE
travis: Enable qt for all jobs

### PR DESCRIPTION
**Pull request by ken2812221**
- If depends build take more than 20 mins, skip Bitcoin Core build to store depends caches and mark it fail. Then restart the job for Bitcoin Core build.
- Enable Qt build for Windows and 32-bit Linux
- Enable wallet for depends x86-64 Linux
- Disable gui tests for Windows since they are not supported

This would be helpful for upgrading Qt (#12971) and protobuf (#13513)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -28,11 +28,11 @@ env:
 # ARM
     - HOST=arm-linux-gnueabihf PACKAGES="g++-arm-linux-gnueabihf" DEP_OPTS="NO_QT=1" GOAL="install" BITCOIN_CONFIG="--enable-glibc-back-compat --enable-reduce-exports"
 # Win32
-    - HOST=i686-w64-mingw32 DPKG_ADD_ARCH="i386" DEP_OPTS="NO_QT=1" PACKAGES="python3 nsis g++-mingw-w64-i686 wine-binfmt wine32" RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-reduce-exports"
+    - HOST=i686-w64-mingw32 DPKG_ADD_ARCH="i386" PACKAGES="python3 nsis g++-mingw-w64-i686 wine-binfmt wine32" RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-reduce-exports --disable-gui-tests"
 # Win64
-    - HOST=x86_64-w64-mingw32 DEP_OPTS="NO_QT=1" PACKAGES="python3 nsis g++-mingw-w64-x86-64 wine-binfmt wine64" RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-reduce-exports"
+    - HOST=x86_64-w64-mingw32 PACKAGES="python3 nsis g++-mingw-w64-x86-64 wine-binfmt wine64" RUN_TESTS=true GOAL="install" BITCOIN_CONFIG="--enable-reduce-exports --disable-gui-tests"
```
Comment (MarcoFalke) [inline .travis.yml]: Might as well change "install" to "deploy", so we get the `setup.exe` as well?

**Comment by MarcoFalke** [comment]
Would be nicer if travis natively supported multi-job-multi-stage builds, but yeah, maybe this hack is required.

File: .travis.yml (hunk 1 above)
Comment (ken2812221) [reply .travis.yml]: Sure

File: .travis.yml [hunk 2]
```diff
@@ -59,20 +59,23 @@ before_script:
...
-    - mkdir build && cd build
-    - DOCKER_EXEC ../configure --cache-file=config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG || ( cat config.log && false)
-    - DOCKER_EXEC make distdir VERSION=$HOST
-    - cd bitcoin-$HOST
-    - DOCKER_EXEC ./configure --cache-file=../config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG || ( cat config.log && false)
-    - DOCKER_EXEC make $MAKEJOBS $GOAL || ( echo "Build failure. Verbose build follows." && DOCKER_EXEC make $GOAL V=1 ; false )
-    - if [ "$RUN_TESTS" = "true" ]; then DOCKER_EXEC LD_LIBRARY_PATH=$TRAVIS_BUILD_DIR/depends/$HOST/lib make $MAKEJOBS check VERBOSE=1; fi
-    - if [ "$TRAVIS_EVENT_TYPE" = "cron" ]; then extended="--extended --exclude feature_pruning,feature_dbcrash"; fi
-    - if [ "$RUN_TESTS" = "true" ]; then DOCKER_EXEC test/functional/test_runner.py --combinedlogslen=4000 --coverage --quiet --failfast ${extended}; fi
+    # Skip Bitcoin Core build if depends build take more than 20 mins.
```
[the remark this answers is not in the document]
Comment (ken2812221) [reply .travis.yml]: It does not make travis green. If `test $SECONDS -lt 1200` has failed, travis would turn out red. But it's a good idea to add error message.

File: .travis.yml [hunk 3]
```diff
@@ -112,20 +108,24 @@ before_script:
...
-    - BEGIN_FOLD distdir; DOCKER_EXEC make distdir VERSION=$HOST; END_FOLD
-    - cd bitcoin-$HOST
-    - BEGIN_FOLD configure; DOCKER_EXEC ./configure --cache-file=../config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG || ( cat config.log && false); END_FOLD
-    - BEGIN_FOLD build; DOCKER_EXEC make $MAKEJOBS $GOAL || ( echo "Build failure. Verbose build follows." && DOCKER_EXEC make $GOAL V=1 ; false ); END_FOLD
-    - if [ "$RUN_TESTS" = "true" ]; then BEGIN_FOLD unit-tests; DOCKER_EXEC LD_LIBRARY_PATH=$TRAVIS_BUILD_DIR/depends/$HOST/lib make $MAKEJOBS check VERBOSE=1; END_FOLD; fi
-    - if [ "$TRAVIS_EVENT_TYPE" = "cron" ]; then extended="--extended --exclude feature_pruning,feature_dbcrash"; fi
-    - if [ "$RUN_TESTS" = "true" ]; then BEGIN_FOLD functional-tests; DOCKER_EXEC test/functional/test_runner.py --combinedlogslen=4000 --coverage --quiet --failfast ${extended}; END_FOLD; fi
+    # Skip Bitcoin Core build if depends build take more than 20 mins.
+    - |
+      ( test $SECONDS -lt 1200 || ( echo Travis early exit to cache current state && false ) ) && (
```
Comment (scravy) [inline .travis.yml]: This could simply be its own line in the `script` section.

I really like the rest of the script to be joined into a single block using the `|` YAML goodness, but putting this check on an individual line (`- ...`) would spare the `) && ( LOTSOFLINES )` awkwardness.


```- test $SECONDS -lt 1200 || ( echo ... ; exit 1 )```

**Comment by ken2812221** [comment]
It's time to make Mac build work on master again.

File: .travis.yml (hunk 3 above)
Comment (scravy) [reply .travis.yml]: As I learned in https://github.com/bitcoin/bitcoin/pull/13863 – if you split the script like that then the exit codes of the commands in each line will not make the script fail, i.e. if the unit tests fail the build will not fail.

File: .travis.yml [hunk 4]
```diff
@@ -123,12 +123,11 @@ jobs:
         RUN_TESTS=true
         GOAL="install"
         BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --enable-glibc-back-compat --enable-reduce-exports --with-gui=qt5 CPPFLAGS=-DDEBUG_LOCKORDER"
-# x86_64 Linux, No wallet
+# x86_64 Linux
```
Comment (scravy) [inline .travis.yml]: Is this change intentional? I find it very handy to have a NO_WALLET build available. It has happened to me before that I changed something in a way which required the wallet in a place which does not require a wallet overall. This particular travis build caught that.

**Review by scravy — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 4 above)
Comment (scravy) [reply .travis.yml]: Sorry, I just noticed your description does mention that. So it is intentional. My comment still stands though, I find it beneficial to have a no-wallet build available. In a perfect world a build would build all possible combinations of build flags, in this imperfect world it can at least check the no-wallet case.

File: .travis.yml (hunk 3 above)
Comment (ken2812221) [reply .travis.yml]: Thanks, I'll work on this after #13836 merged.

File: .travis.yml (hunk 4 above)
Comment (ken2812221) [reply .travis.yml]: Indeed, will undo this.

**Comment by ken2812221** [comment]
This is ready for review. Thanks for #13863, I don't have to do a lot of line changes.

**Comment by MarcoFalke** [comment]
As this is effectively reverting a33b7c9cb545985771d074748c0e368ca2d06702, could the GOALs be changed to `deploy`?

**Comment by ken2812221** [comment]
@MarcoFalke Changed goals to deploy, also revert 3d69853090a9382fc31b1558a74627f61f002584 because of cf01fd6f9c1a31d16884cd1a1a686602b4b47027

File: .travis.yml [hunk 5]
```diff
@@ -133,5 +129,5 @@ jobs:
         OSX_SDK=10.11
         RUN_UNIT_TESTS=false
         RUN_FUNCTIONAL_TESTS=false
-        GOAL="all deploy"
+        GOAL="deploy"
```
Comment (MarcoFalke) [inline .travis.yml]: Why is this change needed? It doesn't seem like the mac cross build should be affected at all?

Comment (ken2812221) [reply .travis.yml]: The origin motivation of 3d69853 is to vompile all binaries, not only qt. But after cf01fd6 has been merged, it would build all binaries even when the target is deploy.

Comment (ken2812221) [reply .travis.yml]: @MarcoFalke Do you think that this change is unnecessary?

File: .travis.yml [hunk 6]
```diff
@@ -34,7 +34,7 @@ install:
 before_script:
   - set -o errexit; source .travis/test_05_before_script.sh
 script:
-  - set -o errexit; source .travis/test_06_script.sh
+  - if [ $SECONDS -gt 1200 ]; then set +o errexit; echo "Travis early exit to cache current state"; false; else set -o errexit; source .travis/test_06_script.sh; fi
```
Comment (Sjors) [inline .travis.yml]: Should this be in the `before_script`?

**Comment by Sjors** [comment]
Concept ACK

**Review by Sjors — COMMENTED** [review]
Concept ACK

File: .travis.yml (hunk 6 above)
Comment (MarcoFalke) [reply .travis.yml]: `before_script` does the depends build, which should be cached, so no?

Comment (Sjors) [reply .travis.yml]: I didn't know chache was discared during a premature exit. If so, then this makes sense.
